### PR TITLE
Feature: Add NRF Consumer support OAuth2

### PR DESCRIPTION
**Pull request by andy89923**
This pull request enables UDM to use NRF services through the OAuth2 service authorization.

## Note
The pull request depends on:
- NRF's PR#27. https://github.com/free5gc/nrf/pull/27

### CODE REVIEW

File: internal/context/context.go [hunk 1]
```diff
@@ -45,11 +45,13 @@ type UDMContext struct {
 	NFDiscoveryClient              *Nnrf_NFDiscovery.APIClient
 	UdmUePool                      sync.Map // map[supi]*UdmUeContext
 	NrfUri                         string
+	NrfCerPem                      string
```
[the remark this answers is not in the document]
Comment (andy89923) [reply internal/context/context.go:48]: If the NRF set ```oauth2 = true```, the UDM would check this parameter. 
```go
if oauth2 && udm_context.GetSelf().NrfCerPem == "" {
    logger.CfgLog.Error("OAuth2 enable but no nrfCerPem provided in config.")
}
```
And this parameter is used for verifying token when UDM as a service producer.

File: internal/sbi/consumer/nf_management.go [hunk 2]
```diff
@@ -69,6 +70,13 @@ func SendRegisterNFInstance(nrfUri, nfInstanceId string, profile models.NfProfil
 			resourceUri := res.Header.Get("Location")
 			resouceNrfUri = resourceUri[:strings.Index(resourceUri, "/nnrf-nfm/")]
 			retrieveNfInstanceId = resourceUri[strings.LastIndex(resourceUri, "/")+1:]
+
+			oauth2 := nf.CustomInfo["oauth2"].(bool)
```
Comment (tim-ywliu) [inline internal/sbi/consumer/nf_management.go:74]:
```
oauth2 := false
if nf.CustomInfo != nil {
    oauth2 = nf.CustomInfo["oauth2"].(bool)
}
```

File: internal/sbi/consumer/nf_management.go [hunk 3]
```diff
@@ -69,6 +70,16 @@ func SendRegisterNFInstance(nrfUri, nfInstanceId string, profile models.NfProfil
 			resourceUri := res.Header.Get("Location")
 			resouceNrfUri = resourceUri[:strings.Index(resourceUri, "/nnrf-nfm/")]
 			retrieveNfInstanceId = resourceUri[strings.LastIndex(resourceUri, "/")+1:]
+
+			oauth2 := false
+			if nf.CustomInfo != nil {
+				oauth2 = nf.CustomInfo["oauth2"].(bool)
```
Comment (tim-ywliu) [inline internal/sbi/consumer/nf_management.go:76]: prevent assertion panic:
```
v, ok := nf.CustomInfo["oauth2"].(bool)
if ok {
    oauth2 = v
}
```

File: pkg/factory/config.go [hunk 4]
```diff
@@ -62,6 +62,7 @@ type Configuration struct {
 	Sbi             *Sbi               `yaml:"sbi,omitempty"  valid:"required"`
 	ServiceNameList []string           `yaml:"serviceNameList,omitempty"  valid:"required"`
 	NrfUri          string             `yaml:"nrfUri,omitempty"  valid:"required, url"`
+	NrfCertPem      string             `yaml:"nrfCertPem,omitempty" valid:"type(string),minstringlength(1),optional"`
```
Comment (tim-ywliu) [inline pkg/factory/config.go:65]:
```
NrfCertPem string `yaml:"nrfCertPem,omitempty" valid:"optional"`
```

File: internal/context/context.go [hunk 5]
```diff
@@ -121,6 +125,9 @@ func InitUdmContext(context *UDMContext) {
 		}
 	}
 	udmContext.NrfUri = configuration.NrfUri
+	if configuration.NrfCertPem != "" {
```
Comment (tim-ywliu) [inline internal/context/context.go:128]: no need this if

**Review by tim-ywliu — APPROVED** [review]

**Pull request merged**